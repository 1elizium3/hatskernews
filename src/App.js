import React, { Component } from 'react';
import axios from 'axios';
import './App.css';

import Clock from './component/clock';
import Search from './component/Search/';
import Table, {ButtonWithLoading} from './component/Table/';

const DEFAULT_QUERY = 'react';
const DEFAULT_HPP = '25';

const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page=';
const PARAM_HPP = 'hitsPerPage=';

class App extends Component {

  _isMounted = false;

  constructor() {
    super();
    this.state = {
      results: null,
      searchKey: '',
      searchTerm: DEFAULT_QUERY,
      error: null,
      isLoading: false,
    };
  
    this.onDismiss = this.onDismiss.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
    this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this);
  };

  needsToSearchTopStories(searchTerm) {
    return !this.state.results[searchTerm];
  };

  setSearchTopStories(result) {
    const { hits, page } = result;
    console.log('result', result)
    this.setState(UpdateSearchTopStoriesState(hits, page));
  };

  fetchSearchTopStories(searchTerm, page = 0) {
    this.setState({ isLoading: true });
    axios
      .get(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}
            &${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
      .then(result => this._isMounted && this.setSearchTopStories(result.data))
      .catch(error => this._isMounted && this.setState({ error }));
  };

  componentDidMount() {
    this._isMounted = true;

    const { searchTerm } = this.state;
    this.setState({
      searchKey: searchTerm
    });
    this.fetchSearchTopStories(searchTerm);
  };

  componentWillUnmount() {
    this._isMounted = false;
  };

  onSearchChange(event) {
    console.log('somethingText')
    this.setState({
      searchTerm: event.target.value
    });
  };

  onSearchSubmit(e) {
    const { searchTerm } = this.state;
    this.setState({
      searchKey: searchTerm
    });

    if (this.fetchSearchTopStories(searchTerm)) {
      this.fetchSearchTopStories(searchTerm);
    };
    
    e.preventDefault();
  };

  onDismiss(id) {
    console.log('onDis')

    const { searchKey, results } = this.state;
    const { hits, page } = results[searchKey];

    const isNotId = (item) => item.objectID !== id;
    const updateHits = hits.filter(isNotId);

    this.setState({
      results: { 
        ...results, 
        [searchKey]: {hits: updateHits, page} 
      }
    });
  };

  render() {
    const { searchTerm, results, searchKey, 
            error, isLoading } = this.state;

    const page = (
      results && results[searchKey] && results[searchKey].page
    ) || 0;

    const list = (
      results && results[searchKey] && results[searchKey].hits
    ) || [];

    if (error) {
      return <p>ERROR</p>
    };

    return (
      <div className="page">
        <Clock />
        <div className="interactions">
          <Search
            value={searchTerm}
            onChange={this.onSearchChange} 
            onSubmit={this.onSearchSubmit}
          >
            Поиск
          </Search>
        </div>

        { error 
          ? <div className="interactions">
              <p>Somethisng went wrong</p>
            </div>  
          : <Table 
              list={list}
              onDismiss={this.onDismiss} 
          />
        } 
        
        <div className="interactions">
          <ButtonWithLoading
            isLoading={isLoading}
            onClick={() => this.fetchSearchTopStories(searchTerm, page + 1)}>
              Больше историй
          </ButtonWithLoading>
        </div>
      </div>
    );
  };
};

export const UpdateSearchTopStoriesState = (hits, page) => (prevState) => {
  const { searchKey, results } = prevState;

  const oldHits = results && results[searchKey] 
    ? results[searchKey].hits 
    : [];

  const updateHits = [ ...oldHits, ...hits];

  return {
    results: {
      ...results,
      [searchKey]: {hits: updateHits, page} 
    },
    isLoading: false
  }
};

export default App;