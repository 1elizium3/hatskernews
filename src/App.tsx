import React from 'react';
import axios from 'axios';
import './App.css';

import Clock from './component/clock';
import Search from './component/Search';
import Table, {ButtonWithLoading} from './component/Table';

const DEFAULT_QUERY = 'react';
const DEFAULT_HPP = '20';

const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page=';
const PARAM_HPP = 'hitsPerPage=';

interface IState {
  results: any,
  searchKey: string;
  searchTerm: string;
  error: any;
  isLoading: boolean;
};

type SearKeyTerm = string;

class App extends React.Component<{}, IState> {

  _isMounted = false;
  
  readonly state: IState = {
    results: null,
    searchKey: '',
    searchTerm: DEFAULT_QUERY,
    error: null,
    isLoading: false,
  };

  // constructor() {
  //   super();
  //   this.state = {
  //     results: null,
  //     searchKey: '',
  //     searchTerm: DEFAULT_QUERY,
  //     error: null,
  //     isLoading: false,
  //   };
  
  //   this.onDismiss = this.onDismiss.bind(this);
  //   this.onSearchChange = this.onSearchChange.bind(this);
  //   this.onSearchSubmit = this.onSearchSubmit.bind(this);
  //   this.setSearchTopStories = this.setSearchTopStories.bind(this);
  //   this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
  //   this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this);
  // };

  needsToSearchTopStories = (searchTerm:SearKeyTerm) => {
    return !this.state.results[searchTerm];
  };

  setSearchTopStories = (result: {hits:any, page:number}) => {
    const { hits, page } = result;
    // console.log('result', result);    
    this.setState(UpdateSearchTopStoriesState(hits, page));
  };

  fetchSearchTopStories = (searchTerm:SearKeyTerm, page = 0) => {
    this.setState({ isLoading: true });
    axios
      .get(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}
            &${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
      .then(result => this._isMounted && this.setSearchTopStories(result.data))
      .catch(error => this._isMounted && this.setState({ error }));
  };

  componentDidMount():void {
    this._isMounted = true;

    const { searchTerm } = this.state;
    
    this.setState({ searchKey: searchTerm });
    this.fetchSearchTopStories(searchTerm);
  };

  componentWillUnmount():void {
    this._isMounted = false;
  };

  onSearchChange = (e: {target: {value:string} }) => {
    this.setState({
      searchTerm: e.target.value
    });
  };
  // Поиск на стороне клиента и сервера
  onSearchSubmit = (e: {preventDefault: () => void}) => {
    const { searchTerm } = this.state;
    this.setState({
      searchKey: searchTerm
    });
    // Добавил проверко на не-null для TS
    if (this.fetchSearchTopStories(searchTerm) !== null) {
      this.fetchSearchTopStories(searchTerm);
    };
    
    e.preventDefault();
  };
  // 
  onDismiss = (id: number) => {
    console.log('onDis')

    const { searchKey, results } = this.state;
    const { hits, page } = results[searchKey];

    const isNotId = (item: { objectID: number }) => item.objectID !== id;
    const updateHits = hits.filter(isNotId);

    this.setState({
      results: { 
        ...results, 
        [searchKey]: {hits: updateHits, page} 
      }
    });
  };

  render() {
    const { searchTerm, results, searchKey, 
            error, isLoading } = this.state;

    const page = (
      results && results[searchKey] && results[searchKey].page
    ) || 0;

    const list = (
      results && results[searchKey] && results[searchKey].hits
    ) || [];

    if (error) {
      return <p>ERROR</p>
    };

    return (
      <div className="page">
        <Clock />
        <div className="interactions">
          <Search
            value={searchTerm}
            onChange={this.onSearchChange} 
            onSubmit={this.onSearchSubmit}
          >
            Поиск
          </Search>
        </div>

        { error 
          ? <div className="interactions">
              <p>Somethisng went wrong</p>
            </div>  
          : <Table 
              list={list}
              onDismiss={this.onDismiss} 
          />
        } 
        
        <div className="interactions">
          <ButtonWithLoading
            isLoading={isLoading}
            onClick={() => this.fetchSearchTopStories(searchTerm, page + 1)}>
              Больше историй
          </ButtonWithLoading>
        </div>
      </div>
    );
  };
};

export const UpdateSearchTopStoriesState = (hits: any, page: number) => (prevState: { searchKey: SearKeyTerm; results: any; }) => {
  const { searchKey, results } = prevState;

  const oldHits = results && results[searchKey] 
    ? results[searchKey].hits 
    : [];

  const updateHits = [ ...oldHits, ...hits];

  return {
    results: {
      ...results,
      [searchKey]: {hits: updateHits, page} 
    },
    isLoading: false
  }
};

export default App;