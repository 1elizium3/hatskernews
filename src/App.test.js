import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Enzyme, { shallow, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import App, { Search, Button, 
              Table, UpdateSearchTopStoriesState } from './App';

Enzyme.configure({ adapter: new Adapter() });

describe('App', () => {

  it('отрисовывает без ошибки', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('есть корректный снимок', () => {
    const component = renderer.create(
      <App />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

describe('Search', () => {

  it('отрисовывает без ошибки', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Search>Поиск</Search>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('есть корректный снимок', () =>{
    const component = renderer.create(
      <Search>Поиск</Search>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

describe('Button', () => {
 
  it('отрисовывает без ошибки', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Button>Give Me More</Button>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('кнопка присутствует', () => {
    const element = shallow(
      <Button>Give Me More</Button>
    );
    expect(element.find('.button-inline'));
  });

  test('есть корректный снимок', () =>{
    const component = renderer.create(
      <Button>Give Me More</Button>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

describe('Table', () => {

  const props = {
    list: [
      {title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y'},
      {title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z'},
    ],
    sortKey: 'TITLE',
    isSortRevers: false
  };

  it('отрисовать без ошибки', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Table {...props} onDismiss={() => {}}/>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('show two items in list', () => {
    const element = shallow(
      <Table {...props} onDismiss={() => {}}/>
    );
    expect(element.find('.table-row').length).toBe(2);
  });

  test('есть корректный снимок', () => {
    const component = renderer.create(
      <Table {...props} onDismiss={() => {}}/>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('UpdateSearchTopStoriesState', () => {

  it('отрисовывает без ошибки', () => {
    const upSerState = renderer.create(
      <UpdateSearchTopStoriesState />
    ).toJSON();
    expect(upSerState).toMatchSnapshot();
  });
});





