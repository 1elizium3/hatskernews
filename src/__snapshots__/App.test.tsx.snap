// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App есть корректный снимок 1`] = `
<div
  className="page"
>
  <div>
    <h2>
      Now is 
      16:09:50
      .
    </h2>
  </div>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value="react"
      />
      <button
        type="submit"
      >
        Поиск
      </button>
    </form>
  </div>
  <div
    className="table"
  >
    <div
      className="table-header"
    >
      <span
        style={
          Object {
            "width": "40%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Заголовок
        </button>
      </span>
      <span
        style={
          Object {
            "width": "30%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Автор
        </button>
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Комментарии
        </button>
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Очки
        </button>
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        Архив
      </span>
    </div>
  </div>
  <div
    className="interactions"
  >
    <div>
      <h1>
        <i
          className="fas fa-yin-yang"
        />
      </h1>
    </div>
  </div>
</div>
`;

exports[`Button есть корректный снимок 1`] = `
<button
  className=""
  onClick={[Function]}
  type="button"
>
  Give Me More
</button>
`;

exports[`Search есть корректный снимок 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    type="text"
    value={true}
  />
  <button
    type="submit"
  >
    Поиск
  </button>
</form>
`;

exports[`Table есть корректный снимок 1`] = `
<div
  className="table"
>
  <div
    className="table-header"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Заголовок
      </button>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Автор
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Комментарии
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Очки
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      Архив
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a>
        1
         
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Отбросить
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a>
        2
         
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Отбросить
      </button>
    </span>
  </div>
</div>
`;
